docs/man/man3/SDCard.3
.TH "SDCard" 3 "Mon Feb 13 2023" "Version 0.1" "ViolaoMagico" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SDCard \- A classe \fBSDCard\fP usada para gerenciar um objeto do tipo SD da biblioteca de SD para Arduino\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <SDCard\&.hpp>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSDCard\fP ()"
.br
.RI "Construtor padrão de um novo cartão SD\&. "
.ti -1c
.RI "vector< string > \fBgetList\fP ()"
.br
.RI "Retorna o vetor com o nome de todos os arquivos no SD\&. "
.ti -1c
.RI "void \fBMakelist\fP ()"
.br
.RI "Preenche o vector filesNames com os nomes dos arquivos no cartão\&. "
.ti -1c
.RI "void \fBprintPosition\fP (std::string pFileName)"
.br
.RI "Printa a posição atual dos motores\&. "
.ti -1c
.RI "string \fBreadFile\fP (int pNumFile, char pNumStroke)"
.br
.RI "Recebe como parametro a posição de uma musica no cartão e o n da batida que deverá ser retornada\&. "
.ti -1c
.RI "void \fBwriteInFile\fP (vector< int > pEnginePos)"
.br
.RI "Escreve a posição dos motores no arquivo de armazenar posição\&. "
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "int \fBSD_SPEED\fP"
.br
.RI "BPM contido no arquivo de cada música\&. "
.ti -1c
.RI "int \fBSD_SUBDIVISION\fP"
.br
.RI "Subdivisão ritmíca de cada música\&. "
.in -1c
.SH "Detailed Description"
.PP 
A classe \fBSDCard\fP usada para gerenciar um objeto do tipo SD da biblioteca de SD para Arduino\&. 


.SH "Constructor & Destructor Documentation"
.PP 
.SS "SDCard::SDCard ()"

.PP
Construtor padrão de um novo cartão SD\&. 
.SH "Member Function Documentation"
.PP 
.SS "vector< string > SDCard::getList ()"

.PP
Retorna o vetor com o nome de todos os arquivos no SD\&. 
.PP
\fBReturns\fP
.RS 4
vector<string> vetor de strings com o nome de todos os arquivos armazenados no SD\&. 
.RE
.PP

.SS "void SDCard::Makelist ()"

.PP
Preenche o vector filesNames com os nomes dos arquivos no cartão\&. 
.SS "void SDCard::printPosition (std::string pFileName)"

.PP
Printa a posição atual dos motores\&. 
.PP
\fBParameters\fP
.RS 4
\fIpFileName\fP string contendo o nome do arquivo onde está armazenado as posições dos motores\&. 
.RE
.PP

.SS "string SDCard::readFile (int pNumFile, char pNumStroke)"

.PP
Recebe como parametro a posição de uma musica no cartão e o n da batida que deverá ser retornada\&. 
.PP
\fBParameters\fP
.RS 4
\fIpNumFile\fP 
.br
\fIpNumStroke\fP 
.RE
.PP
\fBReturns\fP
.RS 4
string 
.RE
.PP

.SS "void SDCard::writeInFile (vector< int > pEnginePos)"

.PP
Escreve a posição dos motores no arquivo de armazenar posição\&. 
.PP
\fBParameters\fP
.RS 4
\fIpEnginePos\fP 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "int SDCard::SD_SPEED"

.PP
BPM contido no arquivo de cada música\&. 
.SS "int SDCard::SD_SUBDIVISION"

.PP
Subdivisão ritmíca de cada música\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for ViolaoMagico from the source code\&.


src/SDCard.cpp
#include "SDCard.hpp"

SDCard::SDCard()
{
    // SD_SPEED = 90;
    SD.begin(5);

    if (!SD.exists("musicas"))
    {

        // Cria o diretório de músicas
        SD.mkdir("/musicas");

        // Essa parte não é essencial. Ela cria um arquivo
        // só pra ler e testar
        File teste = SD.open("/musicas/Teste.txt", FILE_WRITE);

        if (teste)
        {
            teste.println("V<60>");
            teste.println("S<1>");
            teste.println("1<AE  A E  DG Be  D B d s >");
            teste.close();
        }
    }

    Makelist();
}

vector<string> SDCard::getList()
{
    return this->mFilesNames;
}

void SDCard::Makelist()
{
    File root = SD.open("/musicas");
    while (true)
    {
        File entry = root.openNextFile();
        if (!entry)
        {
            break;
        }
        this->mFilesNames.push_back(entry.name());
    }
}

void SDCard::printPosition(std::string pFileName)
{
    const char *c = pFileName.c_str();
    File textFile = SD.open(c);
    if (textFile)
    {
        // Serial.print("Conteudo do arquivo de posição: \n");
        while (textFile.available())
        {
            // Serial.write(textFile.read());
        }
        textFile.close();
    }
    else
    {
        Serial.println("error opening file");
    }
}

string SDCard::readFile(int pNumFile, char pNumStroke)
{
    File text = SD.open("/musicas");
    string verify = "";
    string file = "";
    int goBack = 0;
    for (int i = 0; i < pNumFile; i++)
    {
        File textFile = text.openNextFile();
        if (i == pNumFile - 1)
        {
            if (textFile)
            {
                while (textFile.available())
                {
                    char ch = textFile.read();
                    if (ch == pNumStroke)
                    {
                        ch = textFile.read();
                        if (ch == '<')
                        {
                            while (goBack != 1)
                            {
                                ch = textFile.read();
                                if (ch == '>')
                                {
                                    goBack = 1;
                                }
                                else
                                {
                                    file += ch;
                                }
                            }
                        }
                    }
                    else if (ch == 'V')
                    {
                        string num = "";
                        int vel;
                        ch = textFile.read();
                        if (ch == '<')
                        {
                            for (int i = 0; i < 3; i++)
                            {
                                ch = textFile.read();
                                num += ch;
                            }
                            vel = stoi(num);
                            SD_SPEED = vel;
                        }
                        else if (ch == '>')
                        {
                            break;
                        }
                    }
                    else if (ch == 'S')
                    {
                        string sub = "";
                        int numSubdivision;
                        ch = textFile.read();
                        if (ch == '<')
                        {
                            for (int i = 0; ch != '>'; i++)
                            {
                                ch = textFile.read();
                                sub += ch;
                            }
                        }
                        numSubdivision = stoi(sub);
                        SD_SUBDIVISION = numSubdivision;
                    }
                    else
                    {
                        verify = "";
                    }
                }
                return file;
                textFile.close();
            }
        }
    }
}

void SDCard::writeInFile(vector<int> pEnginePos)
{

    // const char * fileNamePtr = fileName.c_str();

    File textFile = SD.open("/final_position.txt", FILE_WRITE);

    if (textFile)
    {
        // Serial.print("\n Gravando a posição dos motores... \n");
        for (auto &i : pEnginePos)
        {
            textFile.println(i);
        }
        // textFile.println(enginePos);
        textFile.close();
    }
}